File: dsa/graph/has_path.py
def has_path_depth(graph, src, dst):
    if src == dst:
        return True
    for neighbor in graph[src]:
        if has_path_depth(graph, neighbor, dst):
            return True
    return False

def has_path_breadth(graph, src, dst):
    queue = [src]

    while len(queue) > 0:
        current = queue.pop(0)
        if current == dst:
            return True
        for neighbor in graph[current]:
            queue.append(neighbor)
    return False

graph = {
	'f': ['g', 'i'],
	'g': ['h'],
	'h': [],
	'i': ['g', 'k'],
	'j': ['i'],
	'k': []
}

if __name__ == '__main__':
    print(has_path_depth(graph, 'f', 'k'))
    print(has_path_breadth(graph, 'f', 'k'))

File: DSA/graph/hasPath.js
const hasPathDepth = (graph, src, dst) => {
	if (src === dst) return true
	for (let neighbor of graph[src]) {
		if (hasPathDepth(graph, neighbor, dst) === true) {
			return true
		}
	}
	return false
}

const hasPathBreadth = (graph, src, dst) => {
	const queue = [src]
	while (queue.length > 0) {
		const current = queue.shift()
		if (current === dst) return true
		for (let neighbor of graph[current]) {
			queue.push(neighbor)
		}
	}
	return false
}

graph = {
	f: ['g', 'i'],
	g: ['h'],
	h: [],
	i: ['g', 'k'],
	j: ['i'],
	k: []
}

console.log(hasPathDepth(graph, 'f', 'k'))
console.log(hasPathBreadth(graph, 'f', 'k'))
